File: program/tryc/try.py
import os
import sys
import shlex

def run_ls(args):
    """
    列出目录内容
    """
    path = args[1] if len(args) > 1 else "."
    try:
        entries = os.listdir(path)
        for e in entries:
            print(e)
    except OSError as e:
        print(f"ls: {e}", file=sys.stderr)

def run_cd(args):
    """
    改变工作目录
    """
    path = args[1] if len(args) > 1 else os.path.expanduser("~")
    try:
        os.chdir(path)
    except OSError as e:
        print(f"cd: {e}", file=sys.stderr)

def run_pwd(args):
    """
    输出当前工作目录
    """
    print(os.getcwd())

def main_loop():
    """
    主循环，读取用户输入并执行
    """
    while True:
        try:
            # 显示当前目录作为提示符
            prompt = f"{os.getcwd()}$ "
            line = input(prompt)
        except EOFError:
            # Ctrl-D 退出
            print()
            break
        except KeyboardInterrupt:
            # Ctrl-C 取消当前输入，继续循环
            print()
            continue

        if not line.strip():
            continue  # 空行不做任何事

        # 解析命令行（支持空格、引号等）
        try:
            args = shlex.split(line)
        except ValueError as e:
            print(f"解析错误: {e}", file=sys.stderr)
            continue

        cmd = args[0].lower()

        if cmd == "exit":
            break
        elif cmd == "pwd":
            run_pwd(args)
        elif cmd == "cd":
            run_cd(args)
        elif cmd == "ls":
            run_ls(args)
        else:
            print(f"{cmd}: command not found", file=sys.stderr)

if __name__ == "__main__":
    main_loop()

File: program/tryc/try.c
/* minishell.c */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <dirent.h>

#define MAX_LINE 1024

void cmd_pwd() {
    char cwd[PATH_MAX];
    if (getcwd(cwd, sizeof(cwd)) != NULL) {
        puts(cwd);
    } else {
        perror("pwd");
    }
}

void cmd_cd(char *path) {
    if (chdir(path) != 0) {
        perror("cd");
    }
}

void cmd_ls(char *path) {
    DIR *d = opendir(path);
    if (!d) {
        perror("ls");
        return;
    }
    struct dirent *entry;
    while ((entry = readdir(d)) != NULL) {
        printf("%s\n", entry->d_name);
    }
    closedir(d);
}

int main() {
    char line[MAX_LINE];
    while (1) {
        char cwd[PATH_MAX];
        getcwd(cwd, sizeof(cwd));
        printf("%s$ ", cwd);
        if (!fgets(line, sizeof(line), stdin)) break;
        line[strcspn(line, "\n")] = 0;   // 去掉换行
        if (strlen(line) == 0) continue;

        char *args[4];
        int argc = 0;
        char *token = strtok(line, " ");
        while (token && argc < 3) {
            args[argc++] = token;
            token = strtok(NULL, " ");
        }
        args[argc] = NULL;

        if (strcmp(args[0], "exit") == 0) break;
        else if (strcmp(args[0], "pwd") == 0) cmd_pwd();
        else if (strcmp(args[0], "cd") == 0) {
            if (argc > 1) cmd_cd(args[1]);
            else cmd_cd("~");
        }
        else if (strcmp(args[0], "ls") == 0) {
            if (argc > 1) cmd_ls(args[1]);
            else cmd_ls(".");
        }
        else {
            fprintf(stderr, "%s: command not found\n", args[0]);
        }
    }
    return 0;
}
